### DISCUSSION TITLE
What's the render world and why does it exist?

### DISCUSSION THREAD

**Discussion by doonv** [bevyengine/bevy#13494] [Q&A]
I want to create my own renderer using Bevy ECS, and I thought a good starting point would be Bevy's renderer. But one thing that's always confused me is the "render world" that Bevy's renderer uses. From what I've seen it seems to be a separate world that lives entirely on the GPU, and/or is only used by the renderer.

But I don't really understand **why** the render world exists, having a clone of the entire world seems quite inefficient. Are there any other alternatives?

**Comment by torsteingrindvik** [comment]
You might be interested in this: https://bevyengine.org/news/bevy-0-6/#pipelined-rendering-extract-prepare-queue-render

After the main world finishes simulating a frame, whatever is necessary to render it is extracted into the render world (so it's not cloning the _entire_ world).

Rendering that frame then happens at the same time simulation of the next frame is happening, which allows overlapping work.
The alternative would be to interact with the GPU directly in the main world and waiting for results before continuing, which isn't pipelining and reduces throughput.

> it seems to be a separate world that lives entirely on the GPU

I wouldn't say it _lives_ on the GPU, the render world exists to send data to the GPU and to also cache data in order to not send things redundantly, as well as to start render (and compute) passes.
But the business logic of the render world happens on the CPU.

The notion of _worlds_ is interesting generally, I think that in the future we might see more worlds than what we have now which are the _main_ and _render_ worlds.
The pipelining solution didn't need to happen using worlds, but it's a very clean separation and in my opinion helps reduce mental complexity (compared to having render activity interwoven in the same world).
